src/main/java/com/uid2/admin/vertx/service/SharingService.java

```diff
@@ -15,82 +15,227 @@
+        this.keyManager = keyManager;
     }
 
     @Override
     public void setupRoutes(Router router) {
         router.get("/api/sharing/lists").handler(
-                auth.handle(this::handleKeyAclListAll, Role.SHARING_PORTAL)
+                auth.handle(this::handleAllowlistListALL, Role.SHARING_PORTAL)
         );
         router.get("/api/sharing/list/:siteId").handler(
-                auth.handle(this::handleKeyAclList, Role.SHARING_PORTAL)
+                auth.handle(this::handleAllowlistList, Role.SHARING_PORTAL)
         );
         router.post("/api/sharing/list/:siteId").handler(
-                auth.handle(this::handleKeyAclSet, Role.SHARING_PORTAL)
+                auth.handle(this::handleAllowlistSet, Role.SHARING_PORTAL)
         );
+
+        router.get("/api/sharing/keysets").handler(
+                auth.handle(this::handleKeysetListAll, Role.ADMINISTRATOR)
+        );
+        router.post("/api/sharing/keyset").handler(
+                auth.handle(this::handleKeysetSet, Role.ADMINISTRATOR)
+        );
+        router.get("/api/sharing/keyset/:keyset_id").handler(
+                auth.handle(this::handleKeysetList, Role.ADMINISTRATOR)
+        );
+    }
+
+    private void handleKeysetSet(RoutingContext rc) {
+        synchronized (writeLock) {
+           try {
+               keysetProvider.loadContent();
+           } catch (Exception e) {
+               LOGGER.error("Failed to load key acls");
+               rc.fail(500);
+           }
+
+            final JsonObject body = rc.body().asJsonObject();
+
+           final JsonArray whitelist = body.getJsonArray("allowlist");
+           Integer keysetId = body.getInteger("keyset_id");
+           final Integer siteId = body.getInteger("site_id");
+
+           final Map<Integer, Keyset> collection = this.keysetProvider.getSnapshot().getAllKeysets();
+           Keyset keyset = collection.get(keysetId);
+
+           String name;
+
+           if (keyset == null) {
+               keysetId = Collections.max(collection.keySet()) + 1;
+               name = "";
+           } else {
+               keysetId = keyset.getKeysetId();
+               name = keyset.getName();
+           }
+
+           final Set<Integer> newlist = whitelist.stream()
+                   .map(s -> (Integer) s)
+                   .collect(Collectors.toSet());
+
+           final Keyset newKeyset = new Keyset(keysetId, siteId, name,
+                   newlist, Instant.now().getEpochSecond(), true, true);
+
+           collection.put(keysetId, newKeyset);
+           try {
+               storeWriter.upload(collection, null);
+               //Create a new key
+               this.keyManager.addKeysetKey(keysetId);
+           } catch (Exception e) {
+               rc.fail(500, e);
+               return;
+           }
+
+
+           JsonObject jo = new JsonObject();
+           jo.put("site_id", newKeyset.getSiteId());
+           jo.put("keyset_id", newKeyset.getKeysetId());
```

Comment:
should keyset_id go first ?

Reply:
Json is not ordered by spec so should be irrelevant.

On another bit, JsonObject can be more readably defined through:
```
JsonObject jo = JsonObject.of(
                    "name", newKeyset.getName(),
                    "allowlist", newKeyset.getAllowedSites(),
                     ...
            )
```

Would make it more readable than a lot of puts.

Also perhaps can give it a better name than `jo` - eg `responseBody`

Reply:
Might be irrelevant for computers, but not for humans. Eg we wouldn't want every keyset to have elements in a different order

Reply:
I mean even our JSON library is not required to preserve the order in which we've added those properties.

Reply:
How do we have a consistent and sensible order in the current Admin GUI "List Keys"?

Reply:
It doesn't randomize them, just doesn't provide guarantees on order being preserved

Reply:
But yeah - as long as we don't actually rely on that, it's fine